src/ui/public/vis_maps/maps_renderbot.js

```diff
@@ -1,86 +1,274 @@
+      this._kibanaMap = new KibanaMap(containerElement, minMaxZoom);
+      this._kibanaMap.addDrawControl();
+      this._kibanaMap.addFitControl();
+      this._kibanaMap.addLegendControl();
+
+      const autoPrecision = _.get(event, 'chart.geohashGridAgg.params.autoPrecision');
+      let previousPrecision = this._kibanaMap.getAutoPrecision();
+      let precisionChange = false;
+      this._kibanaMap.on('zoomchange', () => {
+        precisionChange = (previousPrecision !== this._kibanaMap.getAutoPrecision());
+        previousPrecision = this._kibanaMap.getAutoPrecision();
+      });
+      this._kibanaMap.on('moveend', () => {
+        this._persistUIStateFromMap();
+      });
+      this._kibanaMap.on('zoomend', () => {
 
-    if (this.mapsData) {
-      this.mapsVis.render(this.mapsData, this.uiState);
+        const isAutoPrecision = _.get(this._chartData, 'geohashGridAgg.params.autoPrecision', true);
+        if (!isAutoPrecision) {
+          return;
+        }
+
+        this._dataDirty = true;
+        if (precisionChange) {
+          courier.fetch();
+        } else {
+          this._recreateGeohashLayer();
+          this._dataDirty = false;
+          this._doRenderComplete();
+        }
+      });
+
+
+      this._kibanaMap.on('drawCreated:rectangle', event => {
+        addSpatialFilter(_.get(this._chartData, 'geohashGridAgg'), 'geo_bounding_box', event.bounds);
+      });
+      this._kibanaMap.on('baseLayer:loaded', () => {
+        this._baseLayerDirty = false;
+        this._doRenderComplete();
+      });
+      this._kibanaMap.on('baseLayer:loading', () => {
+        this._baseLayerDirty = true;
+      });
+    }
+
+    _recreateGeohashLayer() {
+      if (this._geohashLayer) {
+        this._kibanaMap.removeLayer(this._geohashLayer);
+      }
+      if (!this._geohashGeoJson) {
+        return;
+      }
+      const geohashOptions = this._getGeohashOptions();
+      this._geohashLayer = new GeohashLayer(this._chartData.geoJson, geohashOptions, this._kibanaMap.getZoomLevel());
+      this._kibanaMap.addLayer(this._geohashLayer);
     }
-  };
 
-  MapsRenderbot.prototype._getMapsParams = function () {
-    const self = this;
 
-    return _.assign(
-      {},
-      self.vis.type.params.defaults,
-      {
-        type: self.vis.type.name,
-        // Add attribute which determines whether an index is time based or not.
-        hasTimeField: self.vis.indexPattern && self.vis.indexPattern.hasTimeField()
-      },
-      self.vis.params
-    );
-  };
+    /**
+     * called on data change
+     * @param esResponse
+     */
+    render(esResponse) {
+      this._dataDirty = true;
+      this._kibanaMapReady.then(() => {
+        this._chartData = this._buildChartData(esResponse);
+        this._geohashGeoJson = this._chartData.geoJson;
+        this._recreateGeohashLayer();
+        this._useUIState();
+        this._kibanaMap.resize();
+        this._dataDirty = false;
```

Comment:
We have a potential timing issue here.

If we have multiple zoom level changes, and we set `this._dataDirty` to false here after we just triggered another `courier.fetch` in the `zoomend` event handler we could potentially emit the `renderComplete` event prematurely.